plugins/browser-plugin-media-tracking/src/index.ts

```diff
@@ -0,0 +1,348 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import { ready_state_consts, network_state_consts } from './consts';
+import { ControlEvents } from './eventGroups';
+import {
+  timeRangesToObjectArray,
+  textTrackListToJson,
+  isTypeTextTrackEvent,
+  checkPercentBoundryArrayIsValid,
+  isTypeDocumentEvent,
+  isElementFullScreen,
+} from './helperFunctions';
+import { SnowplowMediaEvent } from './snowplowEvents';
+import { DocumentEvent, MediaEvent } from './wgEvents';
+import { MediaProperty, VideoProperty } from './wgProperties';
+import { MediaTrackingConfig, SnowplowData, MediaEventType, HTMLMediaElement } from './types';
+import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
+import { buildSelfDescribingEvent, CommonEventProperties, SelfDescribingJson } from '@snowplow/tracker-core';
+import { MediaElement, VideoElement, MediaPlayerEvent } from './contexts';
+
+declare global {
+  interface HTMLVideoElement {
+    autoPictureInPicture?: boolean;
+    disablePictureInPicture: boolean;
+  }
+  interface HTMLMediaElement {
+    disableRemotePlayback: boolean;
+  }
+  interface HTMLAudioElement {
+    disableRemotePlayback: any;
+  }
+  interface Document {
+    pictureInPictureElement: HTMLMediaElement;
+  }
+}
+
+const _trackers: Record<string, BrowserTracker> = {};
+
+export function MediaTrackingPlugin(): BrowserPlugin {
+  return {
+    activateBrowserPlugin: (tracker: BrowserTracker) => {
+      _trackers[tracker.id] = tracker;
+    },
+  };
+}
+
+export function trackMedia(mediaId: string, config: MediaTrackingConfig = {}): void {
```

Comment:
This was `TrackMedia` in README, but is `trackMedia` here. Although my earlier comment still stands, I think this can have a better name.

Reply:
nit: Personal preference I think from a consistency point of view.

```suggestion
export function trackMedia(mediaId: string, config: MediaTrackingConfig = {}) {
```

Reply:
This function is quite hard to read. the mix of functions, variables and actual logic makes it pretty confusing. 

I'd suggest pulling the functions out (nested functions are never that nice). I think that'd probably solve most issues. Then group the variables and then the logic so they live together in this function.